Update/inline help data flow

retrofox opened the pull request:
#### Changes proposed in this Pull Request

This PR changes the way the `<InlineHelp />` requests the data. Right now, it performs the requests imperatively, executing the `setInlineHelpQuerySearch()` action from the UI. It should be handled by `<QueryInlineHelpSearch />`, the data query component designed for this purpose.
For this, I've added an action used to set the query string from the component/UI. This property is passed to the data query component to finally perform the requests.

#### Testing instructions

Fist at all, check that either `<InlineHelp />` as well as `<HelpSearch />` works in the same way that it rightly works on production.

You can take a look at how the app dispatches the actions while you type in the query string input.

<img width="702" alt="Screen Shot 2020-06-29 at 5 59 21 PM" src="https://user-images.githubusercontent.com/77539/86055708-95a57080-ba32-11ea-9648-bf2e7ab5076c.png">

<img width="666" alt="Screen Shot 2020-06-29 at 5 59 52 PM" src="https://user-images.githubusercontent.com/77539/86055705-950cda00-ba32-11ea-8500-8fb4915153a6.png">

<img width="873" alt="Screen Shot 2020-06-29 at 6 00 20 PM" src="https://user-images.githubusercontent.com/77539/86055695-9211e980-ba32-11ea-9235-988938126d70.png">

Fixes #43624

getdave commented on client/components/data/query-inline-help-search/index.jsx:
```diff
@@ -24,7 +24,7 @@ class QueryInlineHelpSearch extends Component {
 
 	request( props ) {
-		if ( props.requesting || ! props.query ) {
+		if ( props.requesting ) {
```
How come it's safe to remove this now?

retrofox replied:
Because the request action deals when the query is empty. It could be confusing but the best way to understand it (it works to me :-D) is keeping in mind that event when the query is empty, it will return the default results list.

```es6
export function requestInlineHelpSearchResults( searchQuery = '' ) {
	return ( dispatch, getState ) => {
		// ...
		// If the search is empty return contextual results and exist
		// early to avoid unwanted network requests.
		if ( ! searchQuery ) {
			dispatch( {
				type: INLINE_HELP_SEARCH_RESET,
				searchResults: contextualResults,
			} );

			// Exit early
			return;
		}
	// ...
```

If not, it's going to show an empty results box:

<img src="https://user-images.githubusercontent.com/77539/86114975-ab05b380-baa1-11ea-9a87-0dc910e051fd.png" width="400px" />

getdave replied:
Yep ok that makes sense.

getdave commented on client/state/action-types.js:
```diff
@@ -429,6 +429,7 @@ export const INLINE_HELP_SEARCH_RESET = 'INLINE_HELP_SEARCH_RESET';
 export const INLINE_HELP_SEARCH_REQUEST_FAILURE = 'INLINE_HELP_SEARCH_REQUEST_FAILURE';
 export const INLINE_HELP_SEARCH_REQUEST_SUCCESS = 'INLINE_HELP_SEARCH_REQUEST_SUCCESS';
 export const INLINE_HELP_SEARCH_REQUEST_API_RESULTS = 'INLINE_HELP_SEARCH_REQUEST_API_RESULTS';
+export const INLINE_HELP_SET_QUERY_SEARCH = 'INLINE_HELP_SET_QUERY_SEARCH';
```
```suggestion
export const INLINE_HELP_SET_SEARCH_QUERY = 'INLINE_HELP_SET_SEARCH_QUERY';
```

The thing being "set" is the "search query" not the "query search". I'm suggesting this as I think it will read better in English. Feel free to tell me if I've misunderstood.

getdave replied:
Note that the property of the state is also `searchQuery` and not `querySearch` so it makes sense to align with this for consistency.

getdave replied:
THis action is still named incorrectly. It should be `INLINE_HELP_SET_SEARCH_QUERY`. This will bring it into line with the other changes you've made to correct this.

getdave commented on client/state/inline-help/actions.js:
```diff
@@ -22,6 +23,21 @@ import {
+ * @param {string} searchQuery - query store to store.
+ * @returns {Function}            Action thunk.
+ */
+export function setInlineHelpQuerySearch( searchQuery = '' ) {
```
Again we'd need to rename this to 

```suggestion
export function setInlineHelpSearchQuery( searchQuery = '' ) {
```

getdave commented on client/state/inline-help/actions.js:
```diff
@@ -22,6 +23,21 @@ import {
 import 'state/inline-help/init';
 
+/**
+ * Set the query search in the sate tree for the inline help.
```
```suggestion
 * Set the search query in the state tree for the inline help.
```

Also note the spelling fix for "state".

getdave commented on client/state/inline-help/actions.js:
```diff
@@ -22,6 +23,21 @@ import {
+/**
+ * Set the query search in the sate tree for the inline help.
+ *
+ * @param {string} searchQuery - query store to store.
```
```suggestion
 * @param {string} searchQuery - query string to persist.
```

Use of double "store" was a little confusing.

getdave commented on client/state/inline-help/reducer.js:
```diff
@@ -82,6 +83,11 @@ export const search = withoutPersistence(
 					selectedResult: -1,
 					hasAPIResults: false,
 				};
+			case INLINE_HELP_SET_QUERY_SEARCH:
```
I notice that `INLINE_HELP_SET_QUERY_SEARCH` and `INLINE_HELP_SEARCH_REQUEST are identical. 

Why is this? 

Do we want to remove one or the other? It certainly feels like an issue to have two actions for the same action.

retrofox replied:
Yes, it's confusing. It seems that we could remove one of them. I added `INLINE_HELP_SET_QUERY_SEARCH` because that is what the action does. Just set the query which will be used to define the request when it's triggered. It doesn't request anything. Let me dig a little bit on this.

retrofox replied:
Thinking well, in general, we define three actions when the app needs to perform an async request.

* `REQUEST `: the request starts. It's useful to know if it _isRequesting_.
* `REQUEST_SUCCESS`: request ends up successfully.
* `REQUEST_FAILURE`: request ends up in an unexpected way.

So maybe we should consider keeping both actions. `INLINE_HELP_SET_QUERY_SEARCH` to set the query, and `INLINE_HELP_SEARCH_REQUEST` to start to request.

getdave replied:
> Just set the query which will be used to define the request when it's triggered.

Within the `search` portion of state the `INLINE_HELP_SEARCH_REQUEST` is setting the search query text. That's it. 

The "requesting" state is set in the `requesting` reducer and is completely separate from the `search` reducer. This is where the action type `INLINE_HELP_SEARCH_REQUEST` is used.

Therefore, I think we should retain the action type and use it in the `requesting` reducer, but remove it from the `search` reducer as it's not longer needed.

retrofox replied:
> > Just set the query which will be used to define the request when it's triggered.
> 
> Within the `search` portion of state the `INLINE_HELP_SEARCH_REQUEST` is setting the search query text. That's it.

yeah, maybe that's a mistake. This action should be only used to trigger the request, imo.

> The "requesting" state is set in the requesting reducer and is completely separate from the search reducer.

Agree, yes! it's aligned with what I said before.

> Therefore, I think we should retain the action type and use it in the requesting reducer, but remove it from the search reducer as it's not longer needed.

#### In short
* The UI should take over setting the search query. In this case, the `<SearchCad />` sets the query when running the `setInlineHelpSearchQuery()` (change purposed on this PR)
* The action ``requestInlineHelpSearchResults` function which performs the request should not take care of the query but all requesting process: populate items with the endpoint response, handle errors if something fails, reset when there aren't results, etc. Need to keep in an open eye here.
* The connection between the UI and data flow is made by the data query component. 💥 
```es6
<QueryInlineHelpSearch query={ searchQuery } />
```
When the `searchQuery` changes, `<QueryInlineHelpSearch />` performs a new request and the reducer populates the state with the endpoint response.

retrofox commented:
> This is primarily because I'm a little confused as to what is changing and _why_.

:+1:


> We appear to have swapped out the call to `requestInlineHelpSearchResults` with `setInlineHelpQuerySearch` but ultimately they both appear do _exactly_ the same thing in terms of updating the `searchQuery` property in the state.

Yes, as I said in a comment in the code, maybe we should remove one of them. Again, I've added this one because that is what we want: set the search query string when the user types in the input field (UI).

> 
> In `master` we have a separation of concerns with the UI in one place and the data fetching in `< QueryInlineHelpSearch />`. The UI calls out to `requestInlineHelpSearchResults` when it needs results. In this PR we have just changed the intent to be "setting" rather than "requesting" but everything else appears to be the same.

I'm not sure we have a separation of concerns since the component performs the request explicitly (maybe I'm wrong). Let me try to develop the idea

Fisrt, In master, `<InlineHelpSearchCard />` [runs the action explicitly](https://github.com/Automattic/wp-calypso/blob/master/client/blocks/inline-help/inline-help-search-card.jsx#L91) performing a [request to the server](https://github.com/Automattic/wp-calypso/blob/f84a2eb6bca850a2cac957fc908efc5717c52602/client/state/inline-help/actions.js#L34)...

```es6
class InlineHelpSearchCard extends Component {
	// ...
	searchHelperHandler = ( searchQuery ) => {
		// ...

		// Make a search
		this.props.requestInlineHelpSearchResults( searchQuery );
	};
}
```
On this point, we should keep in mind that `<InlineHelpSearchCard />` is a UI component.

Second, at the same time the `<HelpSearchResults />` component [renders `< QueryInlineHelpSearch />`](https://github.com/Automattic/wp-calypso/blob/f84a2eb6bca850a2cac957fc908efc5717c52602/client/blocks/inline-help/inline-help-search-results.jsx#L78), ...

```es6
function HelpSearchResults( props ) {
	// ...
	return (
		<div>
			<QueryInlineHelpSearch query={ searchQuery } requesting={ isSearching } />
			{ renderSearchResults() }
		</div>
	);
}

```

...a data query component that should take over of the requests:

```es6
class QueryInlineHelpSearch extends Component {
	// ...
	request( props ) {
		if ( props.requesting ) {
			return;
		}

		props.requestInlineHelpSearchResults( props.query );
	}

	render() {
		return null;
	}
}
```

So, in short from my point of view, in some way, we are attempting to perform the same request twice. Actually it doesn't happen because it [compares the current state of the query string](https://github.com/Automattic/wp-calypso/blob/f84a2eb6bca850a2cac957fc908efc5717c52602/client/components/data/query-inline-help-search/index.jsx#L19) _before_ to hit the rest API, bit definitely it tries to do it.

> 
> Can you explain the benefits offered by this refactor? How does this improve the data flow? What improves does this provide to us moving forward?

Benefits are avoiding try to request same request twice, improve the abstraction between the UI and the data flow. I've left my explanation above, too 


> 
> I feel like I'm missing something so please excuse my ignorance as I'm aware you've studied this component a lot more thoroughly than I have 🙇

I don't think it's a matter of ignorance but trying to achieve a better approach for this, and I've studied this component _a lot_ neither, so your feedback is more than over welcome.

getdave commented on client/state/inline-help/actions.js:
```diff
@@ -7,6 +7,7 @@ import {
 	INLINE_HELP_SEARCH_REQUEST_FAILURE,
 	INLINE_HELP_SEARCH_REQUEST_SUCCESS,
 	INLINE_HELP_SEARCH_REQUEST_API_RESULTS,
+	INLINE_HELP_SET_QUERY_SEARCH,
```
Action name still incorrect here.

getdave commented on client/state/inline-help/actions.js:
```diff
@@ -22,6 +23,21 @@ import {
+export function setInlineHelpSearchQuery( searchQuery = '' ) {
+	return ( dispatch ) => {
+		dispatch( {
+			type: INLINE_HELP_SET_QUERY_SEARCH,
```
Action name still incorrect here.

getdave commented on client/state/inline-help/reducer.js:
```diff
@@ -10,6 +10,7 @@ import {
 	INLINE_HELP_SELECT_RESULT,
 	INLINE_HELP_SELECT_NEXT_RESULT,
 	INLINE_HELP_SELECT_PREVIOUS_RESULT,
+	INLINE_HELP_SET_QUERY_SEARCH,
```
Action name still incorrect here.

getdave commented on client/state/inline-help/reducer.js:
```diff
@@ -82,6 +83,11 @@ export const search = withoutPersistence(
 					selectedResult: -1,
 					hasAPIResults: false,
 				};
+			case INLINE_HELP_SET_QUERY_SEARCH:
```
Action name still incorrect here.

retrofox replied:
Sorry, something got lost when I rebase. Restoring...

getdave commented:
@retrofox Thanks for taking the time to explain this to me. Makes sense. I responded to your comments above and left some more points (primarily about incorrectly named action type still hanging around).

I think we also need to tidy up the reducers to remove the redundant duplicate action and leave it in place only where it is needed to set the `requesting` state.

> So, in short from my point of view, in some way, we are attempting to perform the same request twice. 

I see this now. Yes, let's move the data fetching to the query component. We can then refactor this to a hook and do away with the query component pattern which is confusing to me.

retrofox commented:
> I think we also need to tidy up the reducers to remove the redundant duplicate action and leave it in place only where it is needed to set the requesting state.

Please in another PR -)

getdave commented:
> > I think we also need to tidy up the reducers to remove the redundant duplicate action and leave it in place only where it is needed to set the requesting state.
> 
> Please in another PR -)

Ok, but we really need to follow up straight away to remove the duplication of `INLINE_HELP_SEARCH_REQUEST` within the `search` reducer in favour of the `INLINE_HELP_SET_SEARCH_QUERY`. It's definitely a warning sign to see different action types causing the same state update to occur.

Once there's an Issue for this we can 👍 this PR.

getdave commented on client/blocks/inline-help/inline-help-search-results.jsx:
```diff
@@ -75,7 +75,7 @@ function HelpSearchResults( {
 
 	return (
 		<div>
-			<QueryInlineHelpSearch query={ searchQuery } requesting={ isSearching } />
+			<QueryInlineHelpSearch query={ searchQuery } />
 			{ renderSearchResults() }
 		</div>
```
We can use a Fragment here `<>` instead of a `<div>`.

retrofox replied:
Glad to update it but honestly this PR isn't taking over of this specific task. Every time that we add new commits we need to pass the tests process, which is great but sometimes it delays the workflow considerably.
Maybe, instead of adding more changes on this PR, we can create a new janitorial one.

retrofox replied:
![image](https://user-images.githubusercontent.com/77539/86246550-8e40ad00-bb81-11ea-8e23-dc4583c5070e.png)

:'(

retrofox commented on client/blocks/inline-help/inline-help-search-card.jsx:
```diff
@@ -87,12 +87,12 @@ class InlineHelpSearchCard extends Component {
 
 	componentDidMount() {
-		this.props.requestInlineHelpSearchResults();
+		this.props.setInlineHelpSearchQuery();
```
> I shouldn't have to know that setInlineHelpSearchQuery has a default argument.

Debatable :-D. Can we pass the empty string straight to the function argument instead of creating a new const?

retrofox replied:
```es6
this.props.setInlineHelpSearchQuery( '' );
```

gwwar commented on client/state/inline-help/actions.js:
```diff
@@ -22,6 +23,21 @@ import {
+ * @param {string} searchQuery - query string to persist.
+ * @returns {Function}            Action thunk.
+ */
+export function setInlineHelpSearchQuery( searchQuery = '' ) {
```
There's no need to wrap this in a thunk. Let's return a simple action object instead. Thunks are harder to inspect and cannot be serialized.

```jsx
return {
		type: INLINE_HELP_SET_SEARCH_QUERY,
		searchQuery,
};
```

gwwar replied:
Note that `requestInlineHelpSearchResults` can also be rewritten to avoid thunks in favor of data-layer handling, but that's a larger refactor that's optional to tackle.

gwwar commented on client/state/inline-help/reducer.js:
```diff
@@ -82,6 +83,11 @@ export const search = withoutPersistence(
+					...state,
+					searchQuery: action.searchQuery,
+				};
 			case INLINE_HELP_SEARCH_REQUEST:
```
WE leave this for a future follow up, but with `INLINE_HELP_SET_SEARCH_QUERY` we can replace instances of `INLINE_HELP_SEARCH_REQUEST` as this appears to do the same thing and setting a query is more descriptive to the app state here.

retrofox replied:
sounds good to me https://github.com/Automattic/wp-calypso/issues/43845

gwwar commented:
Codewise this looks okay to me, though an additional unit test here would be appreciated for the new action/expected reducer behavior. @retrofox is there supposed to be any change in visible behavior to the user here for the Help component?

Note that for testing we can also dispatch events from the console. https://github.com/Automattic/wp-calypso/tree/5fdfb68495e94547e99ba6e9bfe91ace02f7526f/client/state/console-dispatch

gwwar commented:
@retrofox as to your question on whether the current `inlineHelp` shape makes sense for no-results, it can be improved, but I think it's mostly optional. 

For context, currently any no-result matches, store the default results 
<img width="1369" alt="Screen Shot 2020-07-01 at 10 58 29 AM" src="https://user-images.githubusercontent.com/1270189/86276723-42decc80-bb8a-11ea-841c-128fadd4df62.png">

This isn't particularly interesting state wise, and we could achieve the same effect with an empty array, and let the component show some default results.

However before we jump into fixing that, does it even make sense to cache help search results in memory? Eg how often do users re-search for the same help items? 

Tidying state here is really optional since the main trouble we might run into is taking up too much in-memory space if a user searches a lot without refreshing the browser. It's pretty edge-case.

(Note that `console-dispatch` enhancer is available on prod if we add a ?debug query param).

retrofox commented:
> Ok, but we really need to follow up straight away to remove the duplication of `INLINE_HELP_SEARCH_REQUEST` within the `search` reducer in favour of the `INLINE_HELP_SET_SEARCH_QUERY`. It's definitely a warning sign to see different action types causing the same state update to occur.
> 
> Once there's an Issue for this we can 👍 this PR.

🤝  https://github.com/Automattic/wp-calypso/issues/43845 (we should avoid handshaking for now, though 😅 )

retrofox commented:
> @retrofox as to your question on whether the current `inlineHelp` shape makes sense for no-results, it can be improved, but I think it's mostly optional.

Also, maybe the original idea was making the context help (default help items) dynamic?
Anyway, we've been talking with ajax folks and agreed that it isn't a blocker for us so far. Let's keep the same approach and iterate in the future if it turns out mandatory.

Thanks for your review, Kelly. 🙇‍♂️

retrofox commented:
> @retrofox is there supposed to be any change in visible behavior to the user here for the Help component?

No. The visual checking should be _confirm that both components, `<InlineHelp />` and `<HelpSearch />` (customer home) keep working as expected_
